src/arizona_template_scanner.erl

```diff
@@ -0,0 +1,548 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright 2024 Arizona Framework <contact@arizonaframe.work>
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(arizona_template_scanner).
+-moduledoc false.
+
+%% API functions.
+-export([scan/1]). -ignore_xref([scan/1]).
+
+%% Types
+-export_type([token/0]).
+-export_type([location/0]).
+-export_type([line/0]).
+-export_type([column/0]).
+-export_type([result/0]).
+-export_type([error_reason/0]).
+
+-type state() :: #{
+    line => line(),
+    column => column(),
+    first_column => non_neg_integer(),
+    position => non_neg_integer()
+}.
+-opaque token() :: {open_tag, location(), binary()}
+                 | {attr_key, location(), binary()}
+                 | {attr_value, location(), {text, binary()} | {expr, binary()}}
+                 | {bool_attr, location(), binary()}
+                 | {close_tag, location(), {void, boolean()}}
+                 | {closing_tag, location(), binary()}
+                 | {text, location(), binary()}
+                 | {expr, location(), binary()}.
+-opaque location() :: {line(), column()}.
+-opaque line() :: non_neg_integer().
+-opaque column() :: non_neg_integer().
+-opaque result() :: {ok, [token()]} | {error, {error_reason(), location()}}.
+-opaque error_reason() :: badexpr
```

Comment:
I'd not make simple types opaque unless there's a very specific reason for it. Stuff like binary(), atom(), etc.

Reply:
Done in [5a6d9a9](https://github.com/arizona-framework/arizona/pull/143/commits/5a6d9a95e53042650138144c761ba811884903f1)